api/src/server/event.rs

```diff
@@ -0,0 +1,331 @@
+where
+    R: AsyncRead + Send + Unpin,
+    S: AccessModelStore,
+{
+    let mut car = CarReader::new(reader).await?;
+    let event_cid = *car
+        .header()
+        .roots()
+        .first()
+        .ok_or_else(|| anyhow!("car data should have at least one root"))?;
+
+    debug!(%event_cid, "first root cid");
+
+    let mut car_blocks = HashMap::new();
+    while let Some((cid, bytes)) = car.next_block().await? {
+        car_blocks.insert(cid, bytes);
+    }
+    let event_bytes = get_block(&event_cid, &car_blocks, store).await?;
+    let event: unvalidated::Event =
+        serde_ipld_dagcbor::from_slice(&event_bytes).context("decoding event")?;
+    let init_id = match event {
+        unvalidated::Event::Signed(event) => {
+            let link = event
+                .link()
+                .ok_or_else(|| anyhow!("event should have a link"))?;
+
+            let payload_bytes = get_block(&link, &car_blocks, store).await?;
+            let payload: unvalidated::Payload<Ipld> =
+                serde_ipld_dagcbor::from_slice(&payload_bytes).context("decoding payload")?;
+            match payload {
+                unvalidated::Payload::Init(_) => event_cid,
+                unvalidated::Payload::Data(payload) => payload.id(),
+            }
+        }
+        unvalidated::Event::Time(event) => event.id(),
+    };
+
+    let init_bytes = get_block(&init_id, &car_blocks, store).await?;
+    let init_event: unvalidated::Event =
+        serde_ipld_dagcbor::from_slice(&init_bytes).context("decoding init event")?;
+    if let unvalidated::Event::Signed(event) = init_event {
+        let link = event
+            .link()
+            .ok_or_else(|| anyhow!("init event should have a link"))?;
+
+        let payload_bytes = get_block(&link, &car_blocks, store).await?;
+        let payload: unvalidated::Payload<Ipld> =
+            serde_ipld_dagcbor::from_slice(&payload_bytes).context("decoding init payload")?;
+        if let unvalidated::Payload::Init(payload) = payload {
+            Ok(EventId::new(
+                &network,
+                payload.header().sep(),
+                payload.header().model().as_slice(),
+                payload
+                    .header()
+                    .controllers()
+                    .first()
+                    .ok_or_else(|| anyhow!("init header should contain at least one controller"))?,
+                &init_id,
+                &event_cid,
+            ))
+        } else {
+            bail!("init event payload is not well formed")
+        }
+    } else {
+        bail!("init event should be a signed event")
+    }
+}
+
+// Helper function to get the block first from the CAR file data or otherwise from the store.
+async fn get_block<S: AccessModelStore>(
+    cid: &Cid,
+    car_blocks: &HashMap<Cid, Vec<u8>>,
+    store: &S,
+) -> Result<Vec<u8>> {
+    if let Some(bytes) = car_blocks.get(cid) {
+        Ok(bytes.clone())
```

Comment:
is there any way we could move the bytes instead of cloning them?  Just worried that these blocks could be somewhat large in some cases, would be nice to avoid copying them if possible.

Reply:
Not easily, however these bytes come from the CAR file so they are already in memory so we are duplicating them for a short period of time. I expect this is sufficient for now. But if we determine its an issue we can look into reference counting the data or similar.